scripts/release/audit_rosdistro.py

```diff
@@ -0,0 +1,130 @@
+
+# Copyright 2020 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import argparse
+import copy
+import sys
+
+from ros_buildfarm.argument import add_argument_cache_dir
+from ros_buildfarm.argument import add_argument_config_url
+from ros_buildfarm.argument import add_argument_rosdistro_name
+from ros_buildfarm.config import get_index as get_config_index
+from ros_buildfarm.config import get_release_build_files
+from ros_buildfarm.release_job import partition_packages
+from ros_buildfarm.status_page import get_jenkins_job_urls
+
+from rosdistro import get_distribution_file
+from rosdistro import get_distribution_cache
+from rosdistro import get_index
+
+
+def main(argv=sys.argv[1:]):
+    parser = argparse.ArgumentParser(
+        description='Audit the rosdistro for packages failing to build.')
+    add_argument_config_url(parser)
+    add_argument_rosdistro_name(parser)
+    add_argument_cache_dir(parser, '/tmp/package_repo_cache')
+    args = parser.parse_args(argv)
+
+    recommended_action_count = run_audit(args.config_url, args.rosdistro_name, args.cache_dir)
+    return 1 if recommended_action_count else 0
+
+
+def run_audit(config_url, rosdistro_name, cache_dir):
+    config = get_config_index(config_url)
+    index = get_index(config.rosdistro_index_url)
+    dist_file = get_distribution_file(index, rosdistro_name)
+    dist_cache = get_distribution_cache(index, rosdistro_name)
+    build_files = get_release_build_files(config, rosdistro_name)
+    missing_packages = {}
+    for bf_name, bf_value in build_files.items():
+        missing_packages[bf_name] = copy.deepcopy(bf_value.targets)
+        for target in bf_value.get_targets_list():
+            all_pkgs, missing_pkgs = partition_packages(
+                config_url, rosdistro_name, bf_name, target, cache_dir,
+                deduplicate_dependencies=True, dist_cache=dist_cache)
+            missing_packages[bf_name][target] = missing_pkgs
+            if 'all' in missing_packages[bf_name]:
+                missing_packages[bf_name]['all'] &= missing_pkgs
+            else:
+                missing_packages[bf_name]['all'] = missing_pkgs
+
+            if 'all' in missing_packages:
+                missing_packages['all'] &= missing_pkgs
+            else:
+                missing_packages['all'] = missing_pkgs
+
+    recommended_actions = len(missing_packages['all'])
+    print("# Sync preparation report for %s " % rosdistro_name)
```

Comment:
Please use single quote were possible without escaping. The message also has a trailing space.

Same below.